File: Medium/1992. Find All Groups of Farmland/FindAllGroupsOfFarmland.py
'''
LeetCode Medium: 1992. Find All Groups of Farmland
https://leetcode.com/problems/find-all-groups-of-farmland/

Check FindAllGroupsOfFarmland.cpp for brief solution explanation.
'''
class Solution:
    def findFarmland(self, land: List[List[int]]) -> List[List[int]]:
        def check_bounds(r, c, M, N) -> bool:
            return 0<=r<M and 0<=c<N
        M, N = len(land), len(land[0])
        vis = [[False for i in range(N)]for i in range(M)]
        ans = []
        for i in range(M):
            for j in range(N):
                if vis[i][j]==True or land[i][j]==0:
                    continue
                mnr, mnc, mxr, mxc = i, j, i, j
                ti = i
                while check_bounds(ti, j, M, N)==True and land[ti][j]==1:
                    mxr = ti
                    ti+=1
                tj = j
                while check_bounds(i, tj, M, N)==True and land[i][tj]==1:
                    mxc = tj
                    tj+=1
                for r in range(mnr, mxr+1):
                    for c in range(mnc, mxc+1):
                        vis[r][c] = True
                ans.append([mnr, mnc, mxr, mxc])
        return ans

File: Medium/1992. Find All Groups of Farmland/FindAllGroupsOfFarmland.java
/*
LeetCode Medium: 1992. Find All Groups of Farmland
https://leetcode.com/problems/find-all-groups-of-farmland/

Check FindAllGroupsOfFarmland.cpp for brief solution explanation.
*/
class Solution {
    public boolean check_bounds(int r, int c, int M, int N){
        return 0<=r&&r<M&&0<=c&&c<N;
    }
    public int[][] findFarmland(int[][] land) {
        int M = land.length, N = land[0].length;
        boolean[][]vis = new boolean[M][N];
        ArrayList<ArrayList<Integer>>ans_list = new ArrayList<ArrayList<Integer>>();
        for(int i = 0;i<M;i++){
            for(int j = 0;j<N;j++){
                if(vis[i][j]||land[i][j]==0){
                    continue;
                }
                int mnr = i, mnc = j, mxr = i, mxc = j;
                int ti = i;
                while(check_bounds(ti, j, M, N)&&land[ti][j]==1){
                    mxr = ti++;
                }
                int tj = j;
                while(check_bounds(i, tj, M, N)&&land[i][tj]==1){
                    mxc = tj++;
                }
                for(int r = mnr;r<=mxr;r++){
                    for(int c = mnc;c<=mxc;c++){
                        vis[r][c] = true;
                    }
                }
                ArrayList<Integer>t = new ArrayList<Integer>();
                t.add(mnr);
                t.add(mnc);
                t.add(mxr);
                t.add(mxc);
                ans_list.add(t);
            }
        }
        int[][]ans = new int[ans_list.size()][4];
        for(int i = 0;i<ans_list.size();i++){
            for(int j = 0;j<4;j++){
                ans[i][j] = ans_list.get(i).get(j);
            }
        }
        return ans;
    }
}